File: src/main/java/de/pflugradts/pwman3/application/UserInterfaceAdapterPort.java
package de.pflugradts.pwman3.application;

import de.pflugradts.pwman3.domain.model.transfer.Input;
import de.pflugradts.pwman3.domain.model.transfer.Output;
import io.vavr.control.Try;

/**
 * AdapterPort for receiving {@link Input} from and sending {@link Output} to the user.
 */
public interface UserInterfaceAdapterPort {
    Try<Input> receive(Output output);

    default Try<Input> receive() {
        return receive(Output.empty());
    }

    default boolean receiveConfirmation(Output output) {
        return !receive(output)
                .filter(input -> input.getData().isEmpty())
                .filter(input -> input.getCommand().getFirstByte() == 'c')
                .isEmpty();
    }

    Try<Input> receiveSecurely(Output output);

    default Try<Input> receiveSecurely() {
        return receiveSecurely(Output.empty());
    }

    void send(Output output);

    default void sendLineBreak() {
        send(Output.empty());
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/UserInterfaceAdapterPort.kt
package de.pflugradts.passbird.application

import de.pflugradts.passbird.domain.model.transfer.Input
import de.pflugradts.passbird.domain.model.transfer.Output
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.emptyOutput

interface UserInterfaceAdapterPort {
    fun receive(vararg output: Output): Input
    fun receive(): Input = receive(emptyOutput())
    fun receiveSecurely(output: Output): Input
    fun receiveSecurely(): Input = receiveSecurely(emptyOutput())
    fun receiveLineBreakWithin(milliseconds: Long): Boolean
    fun send(vararg output: Output)
    fun startEphemeralLine(output: Output)
    fun updateEphemeralLine(output: Output)
    fun finishEphemeralLine()
    fun sendLineBreak() = send(emptyOutput())
    fun receiveConfirmation(output: Output) =
        receive(output).run { !isEmpty && data.isEmpty && command.size == 1 && command.firstByte == 'c'.code.toByte() }

    fun receiveYes(output: Output) = receive(output).run {
        !isEmpty && data.isEmpty && command.size == 1 &&
            (command.firstByte == 'Y'.code.toByte() || command.firstByte == 'y'.code.toByte())
    }
    fun warningSound()
}
